src/argparse.zig

```diff
@@ -0,0 +1,297 @@
+	return struct {
+		const Self = @This();
+		pub const Args = T;
+
+		pub fn parse(allocator: NeverFailingAllocator, args: []const u8) !ParseResult(T) {
+			const result = _parse(allocator, args);
+			if(callback != null and result == .success) {
+				try callback.?(result.success);
+			}
+			return result;
+		}
+		pub fn _parse(allocator: NeverFailingAllocator, args: []const u8) ParseResult(T) {
+			switch(@typeInfo(T)) {
+				inline .@"struct" => |s| {
+					return parseStruct(s, allocator, args);
+				},
+				inline .@"union" => |u| {
+					if(u.tag_type == null) @compileError("Union must have a tag type");
+					return parseUnion(u, allocator, args);
+				},
+				else => @compileError("Only structs and unions are supported"),
+			}
+		}
+		fn parseStruct(comptime s: std.builtin.Type.Struct, allocator: NeverFailingAllocator, args: []const u8) ParseResult(T) {
+			var result: T = undefined;
+
+			var split = std.mem.splitScalar(u8, args, ' ');
+			var offset: usize = 0;
+			var count: usize = 0;
+
+			inline for(s.fields) |field| {
+				if(!std.mem.eql(u8, field.name, "flags")) {
+					const _arg = split.next();
+					defer {
+						if(_arg) |arg| offset += arg.len + 1;
+						count += 1;
+					}
+
+					@field(result, field.name) = parseArgument(field.type, _arg) catch |err| {
+						const message = std.fmt.allocPrint(allocator.allocator, "Failed to parse argument {} due to error {s} (offset {})", .{count, @errorName(err), offset}) catch unreachable;
+						return .{.failure = .{.message = message}};
+					};
+				}
+			}
+
+			if(split.next() != null) {
+				const message = std.fmt.allocPrint(allocator.allocator, "Too many arguments for command, expected {}", .{count}) catch unreachable;
+				return .{.failure = .{.message = message}};
+			}
+
+			return .{.success = result};
+		}
+		fn parseArgument(comptime Field: type, _arg: ?[]const u8) !Field {
+			switch(@typeInfo(Field)) {
+				inline .optional => |optionalInfo| {
+					if(_arg == null) return null;
+					return try parseArgument(optionalInfo.child, _arg);
+				},
+				else => |fieldInfo| {
+					const arg = _arg orelse return error.MissingArgument;
+					switch(fieldInfo) {
+						inline .@"struct" => {
+							if(!@hasDecl(Field, "parse")) @compileError("Struct must have a parse function");
+							return try @field(Field, "parse")(arg);
+						},
+						inline .@"enum" => return std.meta.stringToEnum(Field, arg) orelse return error.InvalidEnum,
+						inline .float => |floatInfo| return try std.fmt.parseFloat(std.meta.Float(floatInfo.bits), arg),
+						inline .int => |intInfo| return try std.fmt.parseInt(std.meta.Int(intInfo.signedness, intInfo.bits), arg, 0),
+						inline else => |other| @compileError("Unsupported type " ++ @tagName(other)),
+					}
+				},
+			}
+		}
+		fn parseUnion(comptime u: std.builtin.Type.Union, allocator: NeverFailingAllocator, args: []const u8) ParseResult(T) {
+			var failureMessages: ListUnmanaged([]const u8) = .{};
```

Comment:
Since you join them in the end anyways, just use a list of u8 and add a `\n` at the end of all messages.

Reply:
I think I will just have failure contain ListUnmanaged([]const u8), why bother joining, reallocating etc. If you can just print them separately. I will confirm readability of that solution tho